src/entry.c

```diff
@@ -486,144 +502,193 @@ get_list(bitstr_t * bits, int low, int high, const char *names[],
+					break;
+				}
+				if (is_range_term(ch)) {
+					state = R_FINISH;
+					break;
+				}
+				return (EOF);
 
-	Debug(DPARS | DEXT, ("get_range()...entering, exit won't show\n"));
+			case R_STEP:
+				if (get_number(&num3, 0, PPC_NULL, file) != EOF) {
+					state = R_TERMS;
+					break;
+				}
+				return (EOF);
 
-	if (ch == '*') {
-		/* '*' means "first-last" but can still be modified by /step
-		 */
-		num1 = low;
-		num2 = high;
-		ch = get_char(file);
-		if (ch == EOF)
-			return (EOF);
-	}
-	else {
-		ch = get_number(&num1, low, names, ch, file, ",- \t\n");
-		if (ch == EOF)
-			return (EOF);
+			case R_TERMS:
+				if (is_range_term(ch)) {
+					state = R_FINISH;
+					break;
+				}
+				return (EOF);
 
-		if (ch != '-') {
-			/* not a range, it's a single number.
-			 */
-			if (EOF == set_element(bits, low, high, num1)) {
-				unget_char(ch, file);
+			case R_NUM1:
+				if (ch == '-') {
+					state = R_RANGE;
+					break;
+				}
+				if (ch == '~') {
+					state = R_RANDOM;
+					break;
+				}
+				if (is_range_term(ch)) {
+					num2 = num1;
+					state = R_FINISH;
+					break;
+				}
 				return (EOF);
-			}
-			return (ch);
-		}
-		else {
-			/* eat the dash
-			 */
-			ch = get_char(file);
-			if (ch == EOF)
+
+			case R_RANGE:
+				if (get_number(&num2, low, names, file) != EOF) {
+					state = R_RANGE_NUM2;
+					break;
+				}
 				return (EOF);
 
-			/* get the number following the dash
-			 */
-			ch = get_number(&num2, low, names, ch, file, "/, \t\n");
-			if (ch == EOF || num1 > num2)
+			case R_RANGE_NUM2:
+				if (ch == '/') {
+					state = R_STEP;
+					break;
+				}
+				if (is_range_term(ch)) {
+					state = R_FINISH;
+					break;
+				}
 				return (EOF);
-		}
-	}
 
-	/* check for step size
-	 */
-	if (ch == '/') {
-		/* eat the slash
-		 */
-		ch = get_char(file);
-		if (ch == EOF)
-			return (EOF);
+			case R_RANDOM:
+				if (is_range_term(ch)) {
+					num2 = high;
+					state = R_FINISH;
+				}
+				else if (unget_char(ch, file),
+						get_number(&num2, low, names, file) != EOF) {
+					state = R_TERMS;
+				}
+				/* fail if couldn't find match on previous term
+				 * */
+				else
+					return (EOF);
 
-		/* get the step size -- note: we don't pass the
-		 * names here, because the number is not an
-		 * element id, it's a step size.  'low' is
-		 * sent as a 0 since there is no offset either.
-		 */
-		ch = get_number(&num3, 0, PPC_NULL, ch, file, ", \t\n");
-		if (ch == EOF || num3 == 0)
-			return (EOF);
-	}
-	else {
-		/* no step.  default==1.
-		 */
-		num3 = 1;
-	}
+				/* if invalid random range was selected */
+				if (num1 > num2)
+					num1 = num2 = -1;
 
-	/* num1 (through i) will be validated by set_element() below, but num2
-	 * and num3 are merely used as loop condition and increment, and must
-	 * be validated separately.
-	 */
-	if (num2 < low || num2 > high || num3 > high)
+				/* select random number in range <num1, num2>
+				 * */
+				num1 = num2 = random() % (num2 - num1 + 1) + num1;
```

Comment:
This is biased rng. Would  you please consider https://codereview.stackexchange.com/questions/159604/uniform-random-numbers-in-an-integer-interval-in-c

Reply:
I'm not sure that it is necessary to make rng unbiased in this use case. Do you insist on implementing this feature?

Reply:
OK, it's true that the bias won't be big enough to matter given the range is always fairly short in comparison to the random() output range.